File: code/c/rest/server/src/task_repository.c
#include "task_repository.h"
#include <string.h>
#include <stdlib.h>

void repository_init(TaskRepository* repo) {
    memset(repo, 0, sizeof(TaskRepository));
    pthread_mutex_init(&repo->mutex, NULL);
    repository_load_sample_data(repo);
}

void repository_destroy(TaskRepository* repo) {
    pthread_mutex_destroy(&repo->mutex);
}

void repository_load_sample_data(TaskRepository* repo) {
    pthread_mutex_lock(&repo->mutex);
    
    // Sample task 1
    Task task1;
    task_init(&task1, "Implement C REST API");
    strcpy(task1.description, "Build a REST API server using minimal dependencies");
    task1.status = TASK_STATUS_IN_PROGRESS;
    task1.priority = TASK_PRIORITY_HIGH;
    strcpy(task1.tags[0], "c");
    strcpy(task1.tags[1], "rest");
    strcpy(task1.tags[2], "api");
    task1.tag_count = 3;
    strcpy(task1.assigned_to, "systems-team");
    repo->tasks[repo->count++] = task1;
    
    // Sample task 2
    Task task2;
    task_init(&task2, "Add gRPC support");
    strcpy(task2.description, "Implement gRPC server and client for C");
    task2.status = TASK_STATUS_PENDING;
    task2.priority = TASK_PRIORITY_MEDIUM;
    strcpy(task2.tags[0], "c");
    strcpy(task2.tags[1], "grpc");
    strcpy(task2.tags[2], "protobuf");
    task2.tag_count = 3;
    strcpy(task2.assigned_to, "backend-team");
    repo->tasks[repo->count++] = task2;
    
    // Sample task 3
    Task task3;
    task_init(&task3, "Write unit tests");
    strcpy(task3.description, "Add comprehensive test coverage using CUnit");
    task3.status = TASK_STATUS_PENDING;
    task3.priority = TASK_PRIORITY_HIGH;
    strcpy(task3.tags[0], "testing");
    strcpy(task3.tags[1], "quality");
    task3.tag_count = 2;
    strcpy(task3.assigned_to, "qa-team");
    repo->tasks[repo->count++] = task3;
    
    pthread_mutex_unlock(&repo->mutex);
}

bool repository_create_task(TaskRepository* repo, Task* task) {
    bool success = false;
    
    pthread_mutex_lock(&repo->mutex);
    if (repo->count < MAX_TASKS) {
        task_generate_id(task->id);
        task->created_at = time(NULL);
        task->updated_at = task->created_at;
        repo->tasks[repo->count++] = *task;
        success = true;
    }
    pthread_mutex_unlock(&repo->mutex);
    
    return success;
}

Task* repository_get_task(TaskRepository* repo, const char* id) {
    Task* result = NULL;
    
    pthread_mutex_lock(&repo->mutex);
    for (int i = 0; i < repo->count; i++) {
        if (strcmp(repo->tasks[i].id, id) == 0) {
            result = &repo->tasks[i];
            break;
        }
    }
    pthread_mutex_unlock(&repo->mutex);
    
    return result;
}

bool repository_update_task(TaskRepository* repo, const char* id, const Task* task) {
    bool success = false;
    
    pthread_mutex_lock(&repo->mutex);
    for (int i = 0; i < repo->count; i++) {
        if (strcmp(repo->tasks[i].id, id) == 0) {
            Task* existing = &repo->tasks[i];
            
            // Update fields
            if (strlen(task->title) > 0) {
                strcpy(existing->title, task->title);
            }
            if (strlen(task->description) > 0) {
                strcpy(existing->description, task->description);
            }
            existing->status = task->status;
            existing->priority = task->priority;
            
            if (task->tag_count > 0) {
                existing->tag_count = task->tag_count;
                for (int j = 0; j < task->tag_count; j++) {
                    strcpy(existing->tags[j], task->tags[j]);
                }
            }
            
            if (strlen(task->assigned_to) > 0) {
                strcpy(existing->assigned_to, task->assigned_to);
            }
            
            existing->updated_at = time(NULL);
            success = true;
            break;
        }
    }
    pthread_mutex_unlock(&repo->mutex);
    
    return success;
}

bool repository_delete_task(TaskRepository* repo, const char* id) {
    bool success = false;
    
    pthread_mutex_lock(&repo->mutex);
    for (int i = 0; i < repo->count; i++) {
        if (strcmp(repo->tasks[i].id, id) == 0) {
            // Shift remaining tasks
            for (int j = i; j < repo->count - 1; j++) {
                repo->tasks[j] = repo->tasks[j + 1];
            }
            repo->count--;
            success = true;
            break;
        }
    }
    pthread_mutex_unlock(&repo->mutex);
    
    return success;
}

int repository_list_tasks(TaskRepository* repo, Task* result, int max_results,
                         const char* status, const char* assigned_to, 
                         const char** tags, int tag_count) {
    int count = 0;
    
    pthread_mutex_lock(&repo->mutex);
    for (int i = 0; i < repo->count && count < max_results; i++) {
        Task* task = &repo->tasks[i];
        bool include = true;
        
        // Filter by status
        if (status && strlen(status) > 0) {
            if (task->status != task_string_to_status(status)) {
                include = false;
            }
        }
        
        // Filter by assigned_to
        if (include && assigned_to && strlen(assigned_to) > 0) {
            if (strcmp(task->assigned_to, assigned_to) != 0) {
                include = false;
            }
        }
        
        // Filter by tags
        if (include && tags && tag_count > 0) {
            for (int j = 0; j < tag_count; j++) {
                bool has_tag = false;
                for (int k = 0; k < task->tag_count; k++) {
                    if (strcmp(task->tags[k], tags[j]) == 0) {
                        has_tag = true;
                        break;
                    }
                }
                if (!has_tag) {
                    include = false;
                    break;
                }
            }
        }
        
        if (include) {
            task_copy(&result[count], task);
            count++;
        }
    }
    pthread_mutex_unlock(&repo->mutex);
    
    return count;
}

int repository_get_count(TaskRepository* repo) {
    int count;
    pthread_mutex_lock(&repo->mutex);
    count = repo->count;
    pthread_mutex_unlock(&repo->mutex);
    return count;
}

File: code/objective-c/rest/server/src/TaskRepository.m
#import "TaskRepository.h"

@interface TaskRepository ()
@property (nonatomic, strong) NSMutableDictionary<NSString *, Task *> *tasks;
@property (nonatomic, strong) dispatch_queue_t queue;
@end

@implementation TaskRepository

+ (instancetype)sharedRepository {
    static TaskRepository *sharedInstance = nil;
    static dispatch_once_t onceToken;
    dispatch_once(&onceToken, ^{
        sharedInstance = [[self alloc] init];
    });
    return sharedInstance;
}

- (instancetype)init {
    self = [super init];
    if (self) {
        _tasks = [NSMutableDictionary dictionary];
        _queue = dispatch_queue_create("com.taskapi.repository", DISPATCH_QUEUE_CONCURRENT);
        [self loadSampleData];
    }
    return self;
}

- (void)loadSampleData {
    // Sample task 1
    Task *task1 = [[Task alloc] initWithTitle:@"Implement Objective-C REST API"];
    task1.taskDescription = @"Build a REST API server using Foundation framework";
    task1.status = TaskStatusInProgress;
    task1.priority = TaskPriorityHigh;
    task1.tags = @[@"objective-c", @"rest", @"api"];
    task1.assignedTo = @"ios-team";
    
    // Sample task 2
    Task *task2 = [[Task alloc] initWithTitle:@"Add gRPC support"];
    task2.taskDescription = @"Implement gRPC server and client for Objective-C";
    task2.status = TaskStatusPending;
    task2.priority = TaskPriorityMedium;
    task2.tags = @[@"objective-c", @"grpc", @"protobuf"];
    task2.assignedTo = @"backend-team";
    
    // Sample task 3
    Task *task3 = [[Task alloc] initWithTitle:@"Write unit tests"];
    task3.taskDescription = @"Add comprehensive test coverage using XCTest";
    task3.status = TaskStatusPending;
    task3.priority = TaskPriorityHigh;
    task3.tags = @[@"testing", @"quality"];
    task3.assignedTo = @"qa-team";
    
    dispatch_barrier_async(self.queue, ^{
        self.tasks[task1.taskId] = task1;
        self.tasks[task2.taskId] = task2;
        self.tasks[task3.taskId] = task3;
    });
}

- (NSArray<Task *> *)getAllTasks {
    __block NSArray<Task *> *allTasks;
    dispatch_sync(self.queue, ^{
        allTasks = [self.tasks allValues];
    });
    return allTasks;
}

- (NSArray<Task *> *)getTasksWithStatus:(TaskStatus)status {
    __block NSMutableArray<Task *> *filteredTasks = [NSMutableArray array];
    dispatch_sync(self.queue, ^{
        for (Task *task in [self.tasks allValues]) {
            if (task.status == status) {
                [filteredTasks addObject:task];
            }
        }
    });
    return [filteredTasks copy];
}

- (NSArray<Task *> *)getTasksAssignedTo:(NSString *)assignedTo {
    __block NSMutableArray<Task *> *filteredTasks = [NSMutableArray array];
    dispatch_sync(self.queue, ^{
        for (Task *task in [self.tasks allValues]) {
            if ([task.assignedTo isEqualToString:assignedTo]) {
                [filteredTasks addObject:task];
            }
        }
    });
    return [filteredTasks copy];
}

- (NSArray<Task *> *)getTasksWithTags:(NSArray<NSString *> *)tags {
    __block NSMutableArray<Task *> *filteredTasks = [NSMutableArray array];
    dispatch_sync(self.queue, ^{
        for (Task *task in [self.tasks allValues]) {
            BOOL hasAllTags = YES;
            for (NSString *tag in tags) {
                if (![task.tags containsObject:tag]) {
                    hasAllTags = NO;
                    break;
                }
            }
            if (hasAllTags) {
                [filteredTasks addObject:task];
            }
        }
    });
    return [filteredTasks copy];
}

- (Task *)getTaskById:(NSString *)taskId {
    __block Task *task;
    dispatch_sync(self.queue, ^{
        task = self.tasks[taskId];
    });
    return task;
}

- (Task *)createTask:(Task *)task {
    __block Task *createdTask;
    dispatch_barrier_async(self.queue, ^{
        self.tasks[task.taskId] = task;
        createdTask = task;
    });
    return createdTask;
}

- (Task *)updateTask:(NSString *)taskId withRequest:(UpdateTaskRequest *)request {
    __block Task *updatedTask;
    dispatch_barrier_sync(self.queue, ^{
        Task *existing = self.tasks[taskId];
        if (existing) {
            Task *updated = [existing copy];
            
            if (request.title) {
                updated.title = request.title;
            }
            if (request.taskDescription) {
                updated.taskDescription = request.taskDescription;
            }
            if (request.status) {
                updated.status = [request.status integerValue];
            }
            if (request.priority) {
                updated.priority = [request.priority integerValue];
            }
            if (request.tags) {
                updated.tags = request.tags;
            }
            if (request.assignedTo) {
                updated.assignedTo = request.assignedTo;
            }
            
            updated.updatedAt = [NSDate date];
            self.tasks[taskId] = updated;
            updatedTask = updated;
        }
    });
    return updatedTask;
}

- (Task *)updateTaskStatus:(NSString *)taskId status:(TaskStatus)status {
    UpdateTaskRequest *request = [[UpdateTaskRequest alloc] init];
    request.status = @(status);
    return [self updateTask:taskId withRequest:request];
}

- (BOOL)deleteTask:(NSString *)taskId {
    __block BOOL deleted = NO;
    dispatch_barrier_sync(self.queue, ^{
        if (self.tasks[taskId]) {
            [self.tasks removeObjectForKey:taskId];
            deleted = YES;
        }
    });
    return deleted;
}

- (NSDictionary *)listTasksWithStatus:(NSString *)status
                           assignedTo:(NSString *)assignedTo
                                 tags:(NSArray<NSString *> *)tags
                             pageSize:(NSInteger)pageSize
                            pageToken:(NSString *)pageToken
                               sortBy:(NSString *)sortBy
                            sortOrder:(NSString *)sortOrder {
    
    __block NSMutableArray<Task *> *filteredTasks = [NSMutableArray array];
    
    dispatch_sync(self.queue, ^{
        for (Task *task in [self.tasks allValues]) {
            BOOL includeTask = YES;
            
            // Filter by status
            if (status && status.length > 0) {
                TaskStatus taskStatus = [Task stringToStatus:status];
                if (task.status != taskStatus) {
                    includeTask = NO;
                }
            }
            
            // Filter by assigned to
            if (includeTask && assignedTo && assignedTo.length > 0) {
                if (![task.assignedTo isEqualToString:assignedTo]) {
                    includeTask = NO;
                }
            }
            
            // Filter by tags
            if (includeTask && tags && tags.count > 0) {
                for (NSString *tag in tags) {
                    if (![task.tags containsObject:tag]) {
                        includeTask = NO;
                        break;
                    }
                }
            }
            
            if (includeTask) {
                [filteredTasks addObject:task];
            }
        }
    });
    
    // Sort tasks
    NSSortDescriptor *sortDescriptor;
    if ([sortBy isEqualToString:@"title"]) {
        sortDescriptor = [NSSortDescriptor sortDescriptorWithKey:@"title" 
                                                       ascending:[sortOrder isEqualToString:@"asc"]];
    } else if ([sortBy isEqualToString:@"updated_at"]) {
        sortDescriptor = [NSSortDescriptor sortDescriptorWithKey:@"updatedAt" 
                                                       ascending:[sortOrder isEqualToString:@"asc"]];
    } else {
        // Default sort by created_at
        sortDescriptor = [NSSortDescriptor sortDescriptorWithKey:@"createdAt" 
                                                       ascending:[sortOrder isEqualToString:@"asc"]];
    }
    
    NSArray<Task *> *sortedTasks = [filteredTasks sortedArrayUsingDescriptors:@[sortDescriptor]];
    
    // Apply pagination
    NSInteger offset = pageToken ? [pageToken integerValue] : 0;
    NSInteger actualPageSize = (pageSize > 0 && pageSize <= 100) ? pageSize : 20;
    
    NSRange range = NSMakeRange(offset, MIN(actualPageSize, sortedTasks.count - offset));
    NSArray<Task *> *paginatedTasks = [sortedTasks subarrayWithRange:range];
    
    // Prepare response
    NSMutableArray *taskDicts = [NSMutableArray array];
    for (Task *task in paginatedTasks) {
        [taskDicts addObject:[task toDictionary]];
    }
    
    NSString *nextPageToken = nil;
    if (offset + actualPageSize < sortedTasks.count) {
        nextPageToken = [@(offset + actualPageSize) stringValue];
    }
    
    return @{
        @"tasks": taskDicts,
        @"total_count": @(sortedTasks.count),
        @"page_size": @(actualPageSize),
        @"next_page_token": nextPageToken ?: [NSNull null]
    };
}

- (NSInteger)taskCount {
    __block NSInteger count;
    dispatch_sync(self.queue, ^{
        count = self.tasks.count;
    });
    return count;
}

- (void)clearAllTasks {
    dispatch_barrier_async(self.queue, ^{
        [self.tasks removeAllObjects];
    });
}

@end